src/jrd/Optimizer.cpp

```diff
@@ -1003,26 +1006,35 @@ void OptimizerRetrieval::getInversionCandidates(InversionCandidateList* inversio
 				if (segment->scope == scope)
 					scratch.scopeCandidate = true;
 
-				if (segment->scanType != segmentScanMissing && !(scratch.idx->idx_flags & idx_unique))
+				const USHORT iType = scratch.idx->idx_rpt[j].idx_itype;
+
+				if (iType >= idx_first_intl_string)
 				{
 					const USHORT iType = scratch.idx->idx_rpt[j].idx_itype;
+					TextType* textType = INTL_texttype_lookup(tdbb, INTL_INDEX_TO_TEXT(iType));
 
-					if (iType >= idx_first_intl_string)
+					if (segment->scanType != segmentScanMissing && !(scratch.idx->idx_flags & idx_unique))
 					{
-						TextType* textType = INTL_texttype_lookup(tdbb, INTL_INDEX_TO_TEXT(iType));
-
 						if (textType->getFlags() & TEXTTYPE_SEPARATE_UNIQUE)
 						{
 							// ASF: Order is more precise than equivalence class.
 							// We can't use the next segments, and we'll need to use
 							// INTL_KEY_PARTIAL to construct the last segment's key.
-							scratch.fuzzy = true;
+							scratch.usePartialKey = true;
 						}
 					}
+
+					if (segment->scanType == segmentScanStarting)
+					{
+						if (textType->getFlags() & TEXTTYPE_MULTI_STARTING_KEY)
+							scratch.useMultiStartingKeys = true;	// use INTL_KEY_MULTI_STARTING
+						else
+							scratch.usePartialKey = true;
+					}
 				}
 
 				// Check if this is the last usable segment
-				if (!scratch.fuzzy &&
+				if (!scratch.usePartialKey &&
```

Comment:
I believe this check should be extended:
`if (!scratch.usePartialKey && !scratch.useMultiStartingKeys &&`
as I don't see how for index {A, B} segment B can be used if segment A is looked up using irb_multi_starting.

Reply:
I seem to be wrong here, sorry. `usePartialKey` belongs to `segmentScanStarting` and should never appear for equalities.